server/src/main/java/org/apache/druid/discovery/BaseNodeRoleWatcher.java

```diff
@@ -209,13 +221,59 @@ private void removeNode(DiscoveryDruidNode druidNode)
     }
   }
 
+  public void cacheInitializedTimedOut()
+  {
+    synchronized (lock) {
+      LOGGER.warn("Cache for node role [%s] could not be initialized before timeout.", nodeRole.getJsonName());
```

Comment:
This method is scheduled unconditionally for 30s after construction. It shouldn't log any warnings if the cache has already been initialized by then, otherwise people will see these warnings even when nothing has actually timed out.

Reply:
Fixed